generator/src/codegen/cpp.rs

```diff
@@ -0,0 +1,385 @@
+            cg!(self, "return *this;");
+            self.dedent();
+            cg!(self, "}}");
+            cg!(self);
+            cg!(self, "{}& {}::make_{}() {{", type_, name, elem.name());
+            self.indent();
+            cg!(self, "{} tmp;", type_);
+            cg!(self, "set_{}(tmp);", elem.name());
+            cg!(self, "return std::get<static_cast<size_t>({})>(__data);", selection);
+            self.dedent();
+            cg!(self, "}}");
+            cg!(self);
+        }
+        cg!(self, "{0}::Selection {0}::selection() const noexcept {{", name);
+        self.indent();
+        cg!(self, "const size_t index = __data.index();");
+        cg!(self, "return static_cast<Selection>(index);");
+        self.dedent();
+        cg!(self, "}}");
+        cg!(self);
+        Ok(())
+    }
+
+    fn write_choice(&mut self, name: &str, choice: &Choice) -> Result<()> {
+        self.select(Select::H);
+        self.write_choice_header(name, choice)?;
+        self.select(Select::CPP);
+        self.write_choice_source(name, choice)
+    }
+
+    fn write_sequence_header(&mut self, name: &str, s: &Sequence) -> Result<()> {
+        if let Some(doc) = s.doc().as_ref() {
+            cg!(self, "/* {} */", doc);
+        }
+        cg!(self, "class {} {{", name.to_camel_case());
+        self.indent();
+        cg!(self, "public:");
+        self.indent();
+        for elem in s.elements() {
+            let type_ = Generator::get_type(elem);
+            cg!(self, "const {}& get_{}() const noexcept;", type_, elem.name());
+            cg!(self, "{0}& set_{1}(const {2}& {1});", name, elem.name(), type_);
+            cg!(self);
+        }
+        cg!(self);
+        self.dedent();
+        cg!(self, "private:");
+        self.indent();
+        for elem in s.elements() {
+            let type_ = Generator::get_type(elem);
+            if let Some(doc) = elem.doc().as_ref() {
+                cg!(self, "/* {} */", doc);
+            }
+            cg!(self, "{} {};", type_, elem.name());
+        }
+        self.dedent();
+        self.dedent();
+        cg!(self, "}};");
+        Ok(())
+    }
+
+    fn write_sequence_source(&mut self, name: &str, s: &Sequence) -> Result<()> {
+        let name = name.to_camel_case();
+        for elem in s.elements() {
+            let type_ = Generator::get_type(elem);
+            cg!(self, "const {}& {}::get_{}() const noexcept {{", type_, name, elem.name());
+            self.indent();
+            cg!(self, "return {};", elem.name());
+            self.dedent();
+            cg!(self, "}}");
+            cg!(self);
+            cg!(self, "{0}& {0}::set_{1}(const {2}& {1}) {{", name, elem.name(), type_);
+            self.indent();
+            cg!(self, "this->{0} = {0};", elem.name());
+            cg!(self, "return *this;");
+            self.dedent();
+            cg!(self, "}}");
+            cg!(self);
+        }
+        Ok(())
+    }
+
+    #[allow(dead_code)]
```

Comment:
What is this allowing below?

Reply:
This allows for the following function to never be called and you won't get a warning